### PR TITLE
Fast-path some relations via strict equality

### LINKED ISSUE #104583: Exponential compile time with nested `&dyn Fn` type
Found with a [modified fuzz-rustc](https://github.com/jruderman/fuzz-rustc)

### Code

```rust
fn id(
    f: &dyn Fn(u32)
) -> &dyn Fn(
    &dyn Fn(
        &dyn Fn(
            &dyn Fn(&dyn Fn(&dyn Fn(&dyn Fn(&dyn Fn(&dyn Fn(&dyn Fn(&dyn Fn(&dyn Fn(u32)))))))))
        )
    )
) {
    f
}

fn main() {}
```

### Time complexity

Seems exponential:

| Nesting level | Time |
|---------------|------|
| 9  |  0.9 sec |
| 10 |  4.2 sec |
| 11 |  17 sec |
| 12 |  80 sec |

### Where it's slow

* From `-Z time-passes`: `item_bodies_checking` ([full output](https://github.com/rust-lang/rust/files/10044169/ztp.txt))
* From sampling: seems like recursion involving type inference ([head of sample tree](https://github.com/rust-lang/rust/files/10044156/Sample.head.txt))

### Not a recent regression

It's slow at least as far back as nightly-2018-01-01

### Version

```
rustc 1.67.0-nightly (a00f8ba7f 2022-11-15)
binary: rustc
commit-hash: a00f8ba7fcac1b27341679c51bf5a3271fa82df3
commit-date: 2022-11-15
host: x86_64-apple-darwin
release: 1.67.0-nightly
LLVM version: 15.0.4
```

@rustbot label +I-compiletime

### PR DESCRIPTION
**Pull request by compiler-errors**
Some relations can take advantage of strict equality (that is, the `==` operator), which is far cheaper to check in some cases like substs and existential predicate lists. 

This also fixes #104583.

r? @ghost

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/check/dropck.rs [hunk 1]
```diff
@@ -248,6 +248,10 @@ impl<'tcx> TypeRelation<'tcx> for SimpleEqRelation<'tcx> {
         self.param_env
     }
 
+    fn fast_equate_combine(&self) -> bool {
+        true
+    }
+
```
Comment (matthiaskrgr) [inline compiler/rustc_hir_analysis/src/check/dropck.rs:254]: should these be marked `#[inline]`?

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/check/dropck.rs:254]: perhaps that would help... 🤔

**Comment by compiler-errors** [comment]
I'll try one more perf run, but unclear that this is worth the complexity...

@bors try @rust-timer queue

File: compiler/rustc_middle/src/ty/relate.rs [hunk 2]
```diff
@@ -29,7 +29,7 @@ pub trait TypeRelation<'tcx>: Sized {
     fn tag(&self) -> &'static str;
 
     /// Returns whether or not structural equality can be used to fast-path this relation
-    fn fast_equate_combine(&self) -> bool;
+    fn fast_equate(&self) -> bool;
```
Comment (faptc) [inline compiler/rustc_middle/src/ty/relate.rs:32]: Annotate inline this declaration instead?

Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/relate.rs:32]: `#[inline]` means nothing on trait methods without a body.

**Comment by compiler-errors** [comment]
> Is this confirmed to fix https://github.com/rust-lang/rust/issues/104583?

Yes, though I can probably reformulate this to be simpler, given that we don't have a perf win in general. Let me close this and go back to the drawing board.

**Pull request closed without merging**